Optimize iFft

NoCtrlZ opened the pull request:
# Fft Optimization
Hi there.
I optimized `ifft` with more efficient rayon method and `fft` with changing `bit_reverse` algorithm.

## What I did
### replace bitreverse with `u64` reverse_bits
Stop calculate bit reverse and use built-in [method](https://doc.rust-lang.org/std/primitive.u32.html).

### refactoring
Integrate bitreverse and swap into `swap_bit_reverse`.
Modularize butterfly arithmetic to butterfly_arithmetic.
domain::ifft -> arithmetic::best_ifft

- Though
The most efficient is the batched version of fft as [this](https://github.com/appliedzkp/halo2/pull/55/files#diff-1f6174eb1c227f9f0c7bfc68eabc6163cfc87b8cdb6f45fa139ca40ca18049aeR645).
To modularize bitreverse and butterfly arithmetic makes easier to perform as batch.
When we perform batch bitreverse and butterfly arithmetic, we can reuse reversed bit indexes and twiddles as [this](https://github.com/appliedzkp/halo2/pull/55/files#diff-1f6174eb1c227f9f0c7bfc68eabc6163cfc87b8cdb6f45fa139ca40ca18049aeR645) and, we can reduce reverse bit and twiddles calculation.
I think we should separate the structure and algorithm to make it easier to be optimized.
The `ifft` can be independent from `EvaluationDoamin`.

### `scope` to `par_iter`
Replace `scope` with `par_iter`.
The official document says that the `par_iter` is the most efficient basically.
https://docs.rs/rayon/latest/rayon/#how-to-use-rayon

### test and bench
`bit_revsere` and `fft` test.
`ifft` bench.

### adjust multithread overhead
When the k = 8, the overhead of multithread is reasonable so I added condition branch in `multicore`.
The overhead is related task size so I modularize it.

## Bench
The task size condition works fine so lower degree is improved generally.
The `bitreverse` speed is improved so it speeds up 5~10%.

### FFT
| k |  default  | optimize |
| ---- | ---- | ---- |
| 3 | 639.86 ns 642.98 ns 646.28 ns | 639.97 ns 644.09 ns 648.20 ns |
| 4 | 13.713 us 14.033 us 14.386 us | 1.4760 us 1.4853 us 1.4950 us |
| 5 | 21.541 us 21.575 us 21.610 us | 3.2902 us 3.3092 us 3.3286 us |
| 6 | 29.259 us 29.594 us 30.018 us | 7.3762 us 7.3945 us 7.4133 us |
| 7 | 41.503 us 41.589 us 41.683 us | 16.666 us 16.724 us 16.784 us |
| 8 | 61.099 us 62.591 us 64.659 us | 39.766 us 40.486 us 41.447 us |
| 9 | 85.812 us 87.432 us 89.130 us | 82.304 us 83.100 us 84.396 us |
| 10 | 189.75 us 227.31 us 286.96 us | 143.56 us 144.84 us 146.33 us |
| 11 | 307.81 us 320.55 us 334.56 us | 260.60 us 262.47 us 264.41 us |
| 12 | 572.35 us 602.81 us 642.59 us | 516.65 us 522.51 us 527.84 us |
| 13 | 1.0706 ms 1.0997 ms 1.1378 ms | 990.55 us 1.0057 ms 1.0350 ms |
| 14 | 2.2858 ms 2.3534 ms 2.4349 ms | 2.0759 ms 2.1049 ms 2.1372 ms |
| 15 | 4.3920 ms 4.4207 ms 4.4544 ms | 4.1839 ms 4.2064 ms 4.2339 ms |
| 16 | 9.7456 ms 9.9426 ms 10.166 ms | 8.9873 ms 9.1112 ms 9.2438 ms |
| 17 | 21.843 ms 22.563 ms 23.441 ms | 19.496 ms 20.125 ms 20.996 ms |
| 18 | 48.491 ms 50.390 ms 52.815 ms | 44.283 ms 44.972 ms 45.690 ms |

### iFFT

| k |  default  | optimize |
| ---- | ---- | ---- |
3  | 10.892 us 11.390 us 11.899 us | 858.00 ns 862.05 ns 866.08 ns |
4  | 21.124 us 21.426 us 21.792 us | 1.8727 us 1.8796 us 1.8872 us |
5  | 26.511 us 26.724 us 26.978 us | 4.1678 us 4.1850 us 4.2019 us |
6  | 43.917 us 44.081 us 44.268 us | 9.0844 us 9.1195 us 9.1559 us |
7  | 56.390 us 56.661 us 56.933 us | 20.195 us 20.262 us 20.331 us |
8  | 71.472 us 71.831 us 72.229 us | 44.580 us 44.714 us 44.860 us |
9  | 119.10 us 119.37 us 119.68 us | 104.08 us 106.65 us 110.27 us |
10 | 253.42 us 256.95 us 260.37 us | 227.34 us 229.65 us 231.95 us |
11 | 538.85 us 560.10 us 591.15 us | 487.18 us 495.59 us 506.46 us |
12 | 1.0643 ms 1.0856 ms 1.1143 ms | 1.0369 ms 1.0383 ms 1.0400 ms |
13 | 2.2433 ms 2.2986 ms 2.3693 ms | 2.3088 ms 2.3388 ms 2.3844 ms |
14 | 4.6715 ms 4.7023 ms 4.7342 ms | 4.6114 ms 4.6336 ms 4.6570 ms |
15 | 10.286 ms 10.387 ms 10.501 ms | 10.613 ms 10.680 ms 10.756 ms |
16 | 23.683 ms 24.087 ms 24.564 ms | 22.811 ms 23.327 ms 23.977 ms |
17 | 49.733 ms 50.217 ms 50.717 ms | 49.204 ms 49.846 ms 50.753 ms |
18 | 112.70 ms 115.17 ms 118.72 ms | 111.66 ms 113.32 ms 115.38 ms |

I would appreciate it if you would confirm.
Thank you.

str4d commented:
This causes a large regression in `orchard` circuit proving benchmarks. Below measurements are on zcash/orchard@15007026b1181b14f45c5c8cbf32e4fe86bc8d6b with a Ryzen 9 5950X.

## As-is (using `halo2_proofs` from 30f92f3f4b785ea2a32392bf65c1b08f0411567c)
```
proving/bundle/2        time:   [1.1692 s 1.1748 s 1.1801 s]
proving/bundle/3        time:   [1.5929 s 1.6004 s 1.6080 s]
proving/bundle/4        time:   [2.0074 s 2.0169 s 2.0265 s]

verifying/bundle/2      time:   [15.773 ms 15.918 ms 16.066 ms]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
verifying/bundle/3      time:   [17.486 ms 17.675 ms 17.862 ms]
verifying/bundle/4      time:   [19.028 ms 19.228 ms 19.430 ms]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```

## With this PR merged
```
proving/bundle/2        time:   [1.3583 s 1.3702 s 1.3822 s]                              
                        change: [+15.500% +16.626% +17.840%] (p = 0.00 < 0.05)
                        Performance has regressed.
proving/bundle/3        time:   [1.8754 s 1.8885 s 1.9008 s]                              
                        change: [+16.960% +18.005% +18.958%] (p = 0.00 < 0.05)
                        Performance has regressed.
proving/bundle/4        time:   [2.3943 s 2.4119 s 2.4288 s]                              
                        change: [+18.508% +19.583% +20.618%] (p = 0.00 < 0.05)
                        Performance has regressed.

verifying/bundle/2      time:   [15.624 ms 15.786 ms 15.949 ms]                               
                        change: [-2.1847% -0.8251% +0.5663%] (p = 0.24 > 0.05)
                        No change in performance detected.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
verifying/bundle/3      time:   [17.684 ms 17.871 ms 18.052 ms]                               
                        change: [-0.3847% +1.1056% +2.6150%] (p = 0.15 > 0.05)
                        No change in performance detected.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild
verifying/bundle/4      time:   [19.151 ms 19.382 ms 19.624 ms]                               
                        change: [-0.7226% +0.7997% +2.4057%] (p = 0.33 > 0.05)
                        No change in performance detected.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
```

NoCtrlZ commented:
Hi @str4d 
Thank you for check.
In my environment, the result was similar `MacBook Pro (13-inch, 2019, Two Thunderbolt 3 ports)`.
Let me check again.

My diff changes.
```
proving/bundle/1        time:   [2.7179 s 2.8966 s 3.1294 s]                              
                        change: [-6.4165% +3.9583% +15.541%] (p = 0.50 > 0.05)
                        No change in performance detected.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild
Benchmarking proving/bundle/2: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 33.0s.
proving/bundle/2        time:   [2.7541 s 2.8314 s 2.9281 s]                              
                        change: [-3.3454% +2.3964% +8.8485%] (p = 0.46 > 0.05)
                        No change in performance detected.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) low mild
  1 (10.00%) high severe
Benchmarking proving/bundle/3: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 39.2s.
proving/bundle/3        time:   [4.0324 s 4.1709 s 4.3339 s]                              
                        change: [+7.2494% +13.436% +19.976%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild
Benchmarking proving/bundle/4: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 54.8s.
proving/bundle/4        time:   [5.4609 s 5.7325 s 6.0462 s]                              
                        change: [-4.7850% +3.1682% +12.091%] (p = 0.49 > 0.05)
                        No change in performance detected.

verifying/bundle/1      time:   [28.833 ms 29.981 ms 31.410 ms]                               
                        change: [+9.4743% +14.865% +21.604%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  8 (8.00%) high mild
  6 (6.00%) high severe
verifying/bundle/2      time:   [28.157 ms 28.459 ms 28.780 ms]                               
                        change: [+20.751% +22.381% +24.138%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
verifying/bundle/3      time:   [34.423 ms 35.136 ms 35.880 ms]                               
                        change: [+16.022% +20.509% +25.305%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
verifying/bundle/4      time:   [37.928 ms 38.915 ms 40.024 ms]                               
                        change: [+32.275% +35.697% +39.598%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```

NoCtrlZ commented:
I checked `plonk` and `fft` bench.
The `fft` bench is same result as I described but the `plonk` bench slowed down.
I am going to change the algorithm.

NoCtrlZ commented:
# Bench on Github Actions

## Fft

| k | default |
| ---- | ---- |
| 3  |  8.6142 us 8.7320 us 8.8689 us |
| 4  |  12.216 us 12.518 us 12.781 us |
| 5  |  17.907 us 18.075 us 18.248 us |
| 6  |  21.525 us 21.752 us 22.010 us |
| 7  |  32.153 us 32.555 us 33.005 us |
| 8  |  55.390 us 56.248 us 57.279 us |
| 9  |  117.99 us 119.50 us 121.12 us |
| 10 |  225.96 us 230.78 us 236.58 us |
| 11 |  456.56 us 462.97 us 469.76 us |
| 12 |  990.68 us 1.0041 ms 1.0180 ms |
| 13 |  2.0845 ms 2.1082 ms 2.1342 ms |
| 14 |  4.6226 ms 4.6824 ms 4.7474 ms |
| 15 |  9.8348 ms 9.9397 ms 10.050 ms |
| 16 |  22.045 ms 22.281 ms 22.529 ms |
| 17 |  46.408 ms 46.923 ms 47.523 ms |
| 18 |  101.64 ms 102.47 ms 103.32 ms |

## iFft
| k | default |
| ---- | ---- |
| 3  | 15.906 us 16.232 us 16.676 us |
| 4  | 23.633 us 24.137 us 24.568 us |
| 5  | 29.275 us 29.674 us 30.097 us |
| 6  | 35.161 us 35.570 us 36.032 us |
| 7  | 49.167 us 49.667 us 50.194 us |
| 8  | 72.378 us 73.080 us 73.795 us |
| 9  | 127.11 us 129.10 us 131.32 us |
| 10 | 264.60 us 268.29 us 272.14 us |
| 11 | 531.11 us 538.95 us 547.58 us |
| 12 | 1.0617 ms 1.0768 ms 1.0931 ms |
| 13 | 2.3718 ms 2.4091 ms 2.4506 ms |
| 14 | 5.1134 ms 5.1716 ms 5.2331 ms |
| 15 | 11.179 ms 11.294 ms 11.413 ms |
| 16 | 23.664 ms 23.895 ms 24.134 ms |
| 17 | 53.265 ms 53.823 ms 54.415 ms |
| 18 | 111.49 ms 112.73 ms 114.20 ms |

NoCtrlZ commented:
Hi @str4d 
I divided the PR into two parts.
The previous PR is following and it includes refactoring, test and optimization.
https://github.com/zcash/halo2/pull/577
It's so messy.

I am going to create optimization PR after this PR is merged.
I would appreciate it if you would confirm.
Thank you!

daira commented on book/src/design/gadgets/sha256.md:
```diff
@@ -56,7 +56,7 @@ pub trait Sha256Instructions: Chip {
 
 This set of instructions was chosen to strike a balance between the reusability of the
-instructions, and the scope for chips to internally optimise them. In particular, we
+instructions, and the scope for chips to internally optimize them. In particular, we
```
This isn't a typo, it's just the British or New Zealand English spelling. We don't have a preferred English dialect for technical documentation.

NoCtrlZ replied:
Thank you for the review.
Sorry I didn't know that.
I made it back to original spelling.

NoCtrlZ commented on halo2_proofs/src/arithmetic.rs:
```diff
@@ -199,47 +181,40 @@ pub fn best_fft<G: Group>(a: &mut [G], omega: G::Scalar, log_n: u32) {
+pub fn best_ifft<G: Group>(a: &mut [G], omega_inv: G::Scalar, log_n: u32, divisor: G::Scalar) {
+    best_fft(a, omega_inv, log_n);
+    parallelize(a, |a, _| {
+        for a in a {
```
Thank you.
I renamed.

str4d reviewed (changes requested):
LGTM overall, it's a nice refactor.

NoCtrlZ commented:
Hi @str4d 
Thank you for the review.
I modified the PR according to review as following.

> This is only added for use in a test, and should be crate-private unless there is a reason to expose it (in which case it should be documented in the changelog).

It's only for test so I added #[test] flag as you suggested.
> Please squash the typo-related changes from this last commit into the first commit (to remove them from the PR).

I squashed the typo changes.
> Non-blocking: I prefer to avoid wildcard imports.

I avoided the wild card.
> Document this addition in the halo2_proofs changelog:

I added change to changelog.
> These should be proptests, like we do elsewhere in the codebase like

I introduced the proptests to fft test.
> This implicitly relies on usize never exceeding a u64, which there is no guarantee of.

Sorry.
I needed to check it twice.
Thank you for suggestion.
I modified.

I found other section to refactor.
https://github.com/zcash/halo2/compare/df6d307b65e7f2b8ec6082a19a5e381376c999a2..0fc3eba3372f803ab0c87ae7d9764a2e0161075a#diff-e03420d624b59ef5f1ea16419344db88889e1864d7b089db7653ae943b4c7809L230

I would appreciate it if you could confirm.
Thank you!

daira commented on halo2_proofs/src/arithmetic.rs:
```diff
@@ -166,29 +166,11 @@ pub fn best_multiexp<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C]) -> C::Cu
 /// $\omega^{-1}$ in place of $\omega$ and dividing each resulting field element
 /// by $n$.
-///
-/// This will use multithreading if beneficial.
```
It's still the case that this will use multithreading if beneficial, so revert this change.

NoCtrlZ replied:
Thank you.
I modified.
Do you mean we use spawn?
https://doc.rust-lang.org/book/ch16-01-threads.html#creating-a-new-thread-with-spawn

daira commented on halo2_proofs/src/multicore.rs:
```diff
@@ -3,3 +3,14 @@
+        pow += 1;
+    }
+
+    pow
```
This can be written as
```
let num_threads = current_num_threads();
assert!(num_threads > 0 && num_threads <= u64::MAX);
63 - (num_threads as u64).leading_zeros()
```

NoCtrlZ replied:
This is great.

therealyingtong reviewed (approved):
utACK

therealyingtong commented on halo2_proofs/src/arithmetic.rs:
```diff
@@ -411,27 +397,105 @@ pub fn lagrange_interpolate<F: FieldExt>(points: &[F], evals: &[F]) -> Vec<F> {
 
-    let points = (0..5).map(|_| Fp::random(rng)).collect::<Vec<_>>();
-    let evals = (0..5).map(|_| Fp::random(rng)).collect::<Vec<_>>();
+            // This checks whether `best_ifft` is inverse operation of `best_fft`
+            {
+                let domain = EvaluationDomain::<Fp>::new(1, k);
+                let mut a = (0..(1 << k)).map(|_| Fp::random(OsRng)).collect::<Vec<_>>();
+                let b = a.clone();
+                best_fft(&mut a, domain.get_omega(), k);
+                best_ifft(&mut a, domain.get_omega_inv(), k, domain.get_divisor());
+                assert_eq!(a, b);
+            }
```
This should be a separate `test_ifft()`.

therealyingtong commented on halo2_proofs/src/arithmetic.rs:
```diff
@@ -411,27 +397,105 @@ pub fn lagrange_interpolate<F: FieldExt>(points: &[F], evals: &[F]) -> Vec<F> {
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(100))]
+        #[test]
+        fn test_fft(k in 3u32..10) {
```
This proptest should be rewritten to take a single `k: u32`, and sample an arbitrary `a: [Fp: u32]` and `omega: Fp` 

```rust
fn test_fft(k: u32)(
    a in arb_poly(k),
    omega in arb_fp()
)
```
Then, we could run `test_fft(3)`, ..., `test_fft(10)` for the desired range of `k`.
